### PR TITLE
New web3.sha3

### PR DESCRIPTION
**Pull request by carver**
Seeking high-level feedback. I'll figure out the failing tests on Monday.

### What was wrong?

See #288 - new sha3 API, calculated without hitting RPC.

### How was it fixed?

It's not yet. Just sharing the checkpoint where all old tests pass, but not the new ones yet.

Also accidentally included alphabetizing imports in this and another pull, but the merge will get ugly if I try to strip it out. I'll rebase off the other if it gets merged in.

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

File: web3/main.py [hunk 1]
```diff
@@ -152,13 +153,53 @@ def currentProvider(self):
         ))
         return self.manager.providers[0]
 
-    @coerce_return_to_text
-    def sha3(self, value, encoding="hex"):
-        if encoding == 'hex':
-            hex_string = value
+    @staticmethod
+    @apply_to_return_value(encode_hex)
+    def sha3(primitive=None, text=None, hexstr=None, encoding=None):
```
Comment (pipermerriam) [inline web3/main.py]: I think we should either:

* implement validation that ensures that you can't call this function *wrongly* such as `web3.sha3(b'arst', text='bar', hexstr='baz')`
* do some shenanigans with the function signature to make it impossible to call in that manner.

File: web3/main.py [hunk 2]
```diff
@@ -152,13 +153,53 @@ def currentProvider(self):
...
+        if encoding is not None:
+            warnings.warn(DeprecationWarning(
+                "The encoding keyword has been deprecated.  Please update your "
+                "code to use sha3(text='txt'), sha3(hexstr='0x747874'), "
+                "sha3(b'\\x74\\x78\\x74'), or sha3(0x747874)."
+            ))
         else:
-            hex_string = encode_hex(value)
-        return self.manager.request_blocking('web3_sha3', [hex_string])
+            if not isinstance(primitive, (bytes, int, type(None))):
```
Comment (pipermerriam) [inline web3/main.py]: `elif not ....`

Looks like you have one level of unnecessary indentation here.

File: web3/main.py [hunk 3]
```diff
@@ -23,16 +24,18 @@
 from web3.admin import Admin
 from web3.db import Db
 from web3.eth import Eth
+from web3.iban import Iban
```
Comment (pipermerriam) [inline web3/main.py]: Side Note:  I'd like to extract the iban functionality into a stand-alone library and get it out of web3.  It's a format that isn't used or supported very broadly in the ecosystem and while I'm still fine supporting it, I'd rather it live in a different repository.

File: web3/main.py (hunk 1 above)
Comment (carver) [reply web3/main.py]: I'm anti-shenanigan (at least in this context), will validate.

**Pull request merged**